File: src/Flagrum.Core/Graphics/Models/VertexElement.cs
using Flagrum.Core.Serialization.MessagePack;

namespace Flagrum.Core.Graphics.Models;

public class VertexElement : IMessagePackItem
{
    public uint Offset { get; set; }
    public VertexElementSemantic Semantic { get; set; }
    public VertexElementFormat Format { get; set; }

    public void Read(MessagePackReader reader)
    {
        Offset = reader.Read<uint>();
        Semantic = (VertexElementSemantic)reader.Read<string>();
        Format = (VertexElementFormat)reader.Read<int>();
    }

    public void Write(MessagePackWriter writer)
    {
        writer.Write(Offset);
        writer.Write(Semantic.Value);
        writer.Write((int)Format);
    }
}

File: src/Flagrum.Blender/core/graphics/models/vertexelement.py
from dataclasses import dataclass

from .vertexelementformat import VertexElementFormat
from ...serialization.messagepack.reader import MessagePackReader
from ...serialization.messagepack.writer import MessagePackWriter


@dataclass(init=False)
class VertexElement:
    offset: int
    semantic: str
    format: VertexElementFormat

    def read(self, reader: MessagePackReader):
        self.offset = reader.read()
        self.semantic = reader.read()
        self.format = VertexElementFormat(reader.read())

    def write(self, writer: MessagePackWriter):
        writer.write(self.offset)
        writer.write(self.semantic)
        writer.write(self.format)
